### PR TITLE
ARROW-10966: [C++] Use FnOnce for ThreadPool's tasks instead of std::function

### PR DESCRIPTION
**Pull request by michalursa**
FnOnce drops dependencies on invocation and is lighter weight than std::function

### CODE REVIEW

File: cpp/src/arrow/util/thread_pool.cc [hunk 1]
```diff
@@ -84,10 +84,10 @@ static void WorkerLoop(std::shared_ptr<ThreadPool::State> state,
         break;
       }
       {
-        std::function<void()> task = std::move(state->pending_tasks_.front());
+        FnOnce<void()> task = std::move(state->pending_tasks_.front());
         state->pending_tasks_.pop_front();
         lock.unlock();
-        task();
+        std::move(task)();
       }
```
Comment (bkietz) [inline cpp/src/arrow/util/thread_pool.cc:91]: since invoking task will drop its resources, we no longer need to wrap it in an explicit scope

**Comment by bkietz** [comment]
non apache CI: https://github.com/michalursa/arrow/runs/1667448683

Rust failures are unrelated.

**Review by bkietz — APPROVED** [review]
LGTM. Please create a follow up JIRA to apply the same change to TaskGroup

**Comment by michalursa** [comment]
I'll address TaskGroup in [ARROW-11191](https://issues.apache.org/jira/browse/ARROW-11191)

**Pull request closed without merging**